### PR TITLE
fix(gpt-oss): avoid duplicate MoE reduction with DP attention

### LINKED ISSUE #37187: [Bug] GPT-OSS with DP attention enabled produces garbage output (GSM8K 0/128 vs 124/128)
### Checklist

- [x] I searched related issues but found no solution.
- [x] The bug persists in the latest version.
- [x] Issues without environment info and a minimal reproducible demo are hard to resolve and may receive no feedback.
- [x] If this is not a bug report but a general question, please start a discussion at https://github.com/sgl-project/sglang/discussions. Otherwise, it will be closed.
- [x] Please use English. Otherwise, it will be closed.

## Describe the bug

Enabling `--enable-dp-attention` on GPT-OSS-120B (MXFP4) with `--tp 4 --dp 4 --ep 4 --moe-a2a-backend none` completely breaks generation, while the identical command without DP attention works fine.

With DP attention on, responses are dominated by repeated Harmony channel markers, empty final answers, and runaway generation that hits the 8192-token cap. Fixed-prompt evals (temperature 0, seed 4242):

| | GSM8K (first 128, 5-shot) | MMLU (fixed 128 subset) |
|---|---|---|
| DP attention off | 124/128 (96.9%) | 110/128 (85.9%) |
| DP attention on | **0/128 (0%)** | **4/128 (3.1%)** |

The problem is not specific to chunk size, MoE runner backend (`flashinfer_mxfp4` or `triton`), CUDA graph, overlap schedule, radix cache, or hybrid SWA KV pool — we tried toggling each one alone and none recovered basic QA. Also reproducible on current `main` by source inspection; the GPU repro below was on v0.5.18.

### Reproduction

Official image `lmsysorg/sglang:v0.5.18`, model `openai/gpt-oss-120b`:

```bash
python3 -m sglang.launch_server \
  --model-path openai/gpt-oss-120b \
  --tp 4 --dp 4 --ep 4 --moe-a2a-backend none \
  --quantization mxfp4 --dtype bfloat16 \
  --attention-backend trtllm_mha --moe-runner-backend flashinfer_mxfp4 \
  --mem-fraction-static 0.85 --max-total-tokens 1048576 --max-prefill-tokens 131072 \
  --context-length 65536 --disable-prefill-cuda-graph --disable-flashinfer-autotune \
  --random-seed 4242 --cuda-graph-max-bs 128 --max-running-requests 512 \
  --disable-radix-cache --reasoning-parser gpt-oss --tool-call-parser gpt-oss \
  --chunked-prefill-size 65536 --enable-dp-attention
```

The control (working) run is the same command with `--dp 4 --enable-dp-attention` removed and `--chunked-prefill-size 16384` (both land at an effective chunk of 16384 per rank). Eval via the OpenAI-compatible API with temperature 0, top_p 1, seed 4242, `reasoning_effort=low`, max_tokens 8192, concurrency 128; ask something trivial like "What is 2+2? Reply with only the number." — the DP-attention server returns channel-marker gibberish or an empty final answer.

### Environment

Run inside the official Docker image `lmsysorg/sglang:v0.5.18` (ARM64, index digest `sha256:e42d0524...`), so the image itself pins the environment:

- sglang 0.5.18 (launch_server reports v0.5.18; container source matches the `v0.5.18` tag, commit `71de97b2`)
- PyTorch 2.13.0+cu130, FlashInfer 0.6.17, Transformers 5.12.1, Triton 3.7.1, numpy 2.3.5
- 4× NVIDIA GB200 (185 GiB visible per GPU), driver 580.95.05

### PR DESCRIPTION
**Pull request by sghhhh**
## Motivation

Fixes #37187.

Enabling DP attention for GPT-OSS-120B with TP4/DP4/EP4 caused severe output corruption:

| Configuration | GSM8K first 128 | MMLU fixed 128 subset |
| --- | ---: | ---: |
| DP attention disabled | 124/128 | 110/128 |
| DP attention enabled | 0/128 | 4/128 |

Affected responses contained repeated Harmony channel markers, empty final
answers, and generations that frequently reached the 8192-token limit.

### Root cause

With MoE A2A disabled and `TP == attention DP == EP`, every TP rank initially
produces an expert partial `m_i`.

`GptOssSparseMoeBlock.forward_normal()` first performed a TP all-reduce, so
every rank held:

```text
s = m0 + m1 + ... + m(TP-1)
```

DP-attention postprocessing then called `reduce_scatterv` over the same TP
group. Since reduce-scatter also sums its inputs, it combined TP replicated
copies of `s`, producing approximately:

```text
TP * s
```

On TP4, the MoE output was therefore scaled by approximately 4x.

The residual branch was not scaled, so the following residual addition and
normalization could not cancel this error.

### Layer-level evidence

Using the same 80-token input on GPT-OSS-120B MXFP4 with TP4/EP4:

| First-layer boundary | DPA compared with healthy TP reference |
| --- | ---: |
| Input IDs, positions, embeddings, attention-norm input | Exact match |
| Reconstructed Q/K/V and attention-kernel output | Exact match |
| O-projection output | Relative L2 0.002145 |
| MoE input | Relative L2 0.003450 |
| MoE output RMS | 0.803564 -> 3.216738 |
| MoE output after multiplying DPA by 0.25 | Relative L2 0.021860 |
| Next attention-norm input | Relative L2 0.491051 |

The fitted DPA correction factor was `0.249747`, consistent with a duplicate
reduction on TP4.

## Modifications

- Make GPT-OSS use the centralized
  `should_skip_post_experts_all_reduce(is_tp_path=True)` ownership predicate.
- When DP-attention postprocessing owns the reduction, pass the original expert
  partials directly to `reduce_scatterv`.
- Preserve the existing model-side all-reduce for configurations where no
  downstream component owns the reduction.
- Add CPU regression tests covering:
  - TP4/DP4/EP4 with DPA: skip the model-side all-reduce
  - plain TP: preserve the all-reduce
  - fused MLP all-reduce: preserve the existing skip behavior
  - partial attention DP (`TP != attention DP`), including TP4/DP2/EP4:
    preserve all-reduce + scatter

This does not divide the output by TP size and does not modify the
reduce-scatter implementation.

### Scope and Limitations

This PR changes only GPT-OSS post-experts collective ownership. It does not
modify:

- attention or MoE kernels
- quantization
- KV-cache management
- parsers
- CUDA Graph
- overlap scheduling
- chunking

The GPU validation does not cover dynamic DWDP, other TP/DP/EP sizes, full
benchmark datasets, long-context retrieval, or post-fix throughput.

### Related Work

#33135 contains the same centralized GPT-OSS guard as part of a broader draft
FlashInfer A2A/routed-MoE change. This PR isolates the correctness fix and adds
the DP-attention reproduction, layer-level diagnosis, accuracy validation, and
focused regression coverage.

## Accuracy Tests

Validation environment:

- Official `lmsysorg/sglang:v0.5.18`
- `openai/gpt-oss-120b`, MXFP4 weights with BF16 activations
- 4x NVIDIA GB200
- TP4/DP4/EP4
- `trtllm_mha` attention
- `flashinfer_mxfp4` MoE runner
- MoE A2A backend `none`

### Exact same-configuration smoke test

The launch parameters and effective chunk size were unchanged; the only
inference-code change was the post-experts reduction guard.

| Test | Before | After |
| --- | ---: | ---: |
| Basic questions | 1/4 | 4/4 |
| Tool-call structure | Not run after failed smoke | Passed |

### Fixed-prompt accuracy

| Configuration | Configured chunked prefill | Effective chunk/DP rank | GSM8K first 128 | MMLU fixed 128 subset |
| --- | ---: | ---: | ---: | ---: |
| DPA disabled reference | 16384 | 16384 | 124/128 | 110/128 |
| DPA enabled, original code | 16384 | 4096 | 0/128 | 4/128 |
| DPA enabled, fixed | 16384 | 4096 | 121/128 | 110/128 |

All 256 responses from the fixed run had:

- zero empty final answers
- zero 8192-token truncations
- zero API errors
- zero retry errors

All three configurations used `--chunked-prefill-size 16384`. With DP
attention enabled at DP4, the effective chunk size is 4096 tokens per DP
rank (`16384 / 4`), for both the original and fixed runs. The DPA-disabled
reference uses the full 16384-token chunk. The original and fixed DPA runs
therefore used the same chunk-size configuration.

The 256-question run demonstrates recovery to the healthy reference range,
not bitwise equivalence.

The subsets were:

- GSM8K: first 128 test questions with five-shot train examples
- MMLU: 100 `abstract_algebra` and 28 `anatomy` questions

They are diagnostic fixed subsets, not full benchmark scores.

### Unit tests and checks

- New GPT-OSS collective-guard CPU tests: 5/5 passed
- Related runtime-context helper tests: 2/2 passed
- `git diff --check`
- `py_compile`
- Black formatting check
- isort check
- PR Ruff rules
- registered-test CI validation

## Speed Tests and Profiling

No post-fix performance benchmark was run.

The change removes a redundant TP all-reduce from the affected path, but this
PR makes no performance claim. Previous throughput measurements were collected
before the correctness fix and are not used as post-fix performance evidence.

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [x] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations). No documentation change is required for this model correctness fix.
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed). Accuracy results are provided above; no post-fix speed claim is made.
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

---
### CI States

Latest PR Test (Base): :white_check_mark: [Run #34040611753](https://github.com/sgl-project/sglang/actions/runs/34040611753)
Latest PR Test (Extra): :x: [Run #34040611603](https://github.com/sgl-project/sglang/actions/runs/34040611603)
Latest PR Test (AMD ROCm 7.2): :hourglass_flowing_sand: [Run #34040611736](https://github.com/sgl-project/sglang/actions/runs/34040611736)

### CODE REVIEW

**Comment by sghhhh** [comment]
Hi @Fridge003 and @ch-wan, when you have a chance, would you mind taking a look at this GPT-OSS correctness fix?

  This PR addresses the DP-attention double reduction reported in #37187, with
  layer-level root-cause analysis, accuracy validation, and focused regression
  tests.

  I would especially appreciate your feedback on whether the collective
  ownership logic and the scope of the fix look appropriate, and whether there
  are any additional configurations, tests, or validation results that would be
  helpful before merging.

  Thank you very much for your time!

**Comment by sghhhh** [comment]
Hi @Fridge003 @ch-wan, could you please take a look at this straightforward correctness fix? It prevents GPT-OSS from reducing the MoE output twice with DP attention. The reproduction and regression tests are included in the PR. Thanks!

File: python/sglang/srt/models/gpt_oss.py [hunk 1]
```diff
@@ -329,7 +332,15 @@ def forward_normal(
             topk_output = self.topk(router_input, router_logits)
             final_hidden_states = self.experts(hidden_states, topk_output)
 
-        if self.tp_size > 1 and not get_forward().fuse_mlp_allreduce:
+        # With DP attention and a2a none, postprocess_layer combines the
```
Comment (mmangkad) [inline python/sglang/srt/models/gpt_oss.py]: Trim this

File: python/sglang/srt/models/gpt_oss.py [hunk 2]
```diff
@@ -329,7 +332,15 @@ def forward_normal(
...
+        # expert partials with reduce_scatterv (when TP == attn DP == EP).
+        # All-reducing here first would make that sum count every rank's
+        # contribution once more per rank -- e.g. a ~4x-scaled MoE output on
+        # TP4/DP4/EP4. Use the centralized ownership predicate so exactly one
+        # component performs the post-experts reduction.
+        if self.tp_size > 1 and not should_skip_post_experts_all_reduce(
```
Comment (mmangkad) [inline python/sglang/srt/models/gpt_oss.py:335]: Maybe publish `mlp_reduce_scatter` too like the other models

File: test/registered/unit/models/test_gpt_oss_moe_allreduce_guard.py
Comment (mmangkad) [inline test/registered/unit/models/test_gpt_oss_moe_allreduce_guard.py:1]: This is useless, we could add a DPA variant to the gpt-oss 4-gpu e2e test instead

Comment (sghhhh) [reply test/registered/unit/models/test_gpt_oss_moe_allreduce_guard.py:1]: Updated as suggested, thanks!

**Comment by mmangkad** [comment]
/rerun-test test/registered/models_e2e/test_gpt_oss_4gpu_mxfp4.py

File: test/registered/models_e2e/test_gpt_oss_4gpu_mxfp4.py [hunk 3]
```diff
@@ -3,8 +3,8 @@
 from sglang.test.ci.ci_register import register_cuda_ci
 from sglang.test.gpt_oss_common import BaseTestGptOss
 
-register_cuda_ci(est_time=220, stage="base-c", runner_config="4-gpu-h100")
-register_cuda_ci(est_time=220, stage="base-c", runner_config="4-gpu-b200")
+register_cuda_ci(est_time=440, stage="base-c", runner_config="4-gpu-h100")
+register_cuda_ci(est_time=440, stage="base-c", runner_config="4-gpu-b200")
```
Comment (mmangkad) [inline test/registered/models_e2e/test_gpt_oss_4gpu_mxfp4.py]:
```suggestion
register_cuda_ci(est_time=300, stage="base-c", runner_config="4-gpu-h100")
register_cuda_ci(est_time=300, stage="base-c", runner_config="4-gpu-b200")
```

**Review by mmangkad — APPROVED** [review]
